python/lsst/image_cutout_backend/_image_cutout_backend.py

```diff
@@ -0,0 +1,412 @@
+    """Additional FITS metadata about the cutout process.
+
+    This should be merged with ``cutout.getMetadata()`` on write, for types
+    that carry their own metadata.
+    """
+
+    origin_ref: DatasetRef
+    """Fully-resolved reference to the dataset the cutout is from.
+    """
+
+    def mask(self, name: str = "STENCIL") -> None:
+        """Set the bitmask to show the approximate coverage of nonrectangular
+        stencils.
+
+        Parameters
+        ----------
+        name : `str`, optional
+            Name of the mask plane to add and set.
+
+        Notes
+        -----
+        This method modifies `cutout` in place if it is a `Mask`,
+        `MaskedImage`, or `Exposure`.  It does nothing if `cutout` is an
+        `Image`.
+        """
+        if isinstance(self.cutout, Exposure):
+            mask = self.cutout.mask
+        elif isinstance(self.cutout, MaskedImage):
+            mask = self.cutout.mask
+        elif isinstance(self.cutout, Mask):
+            mask = self.cutout
+        else:
+            return
+        mask.addMaskPlane(name)
+        bits = mask.getPlaneBitMask(name)
+        self.pixel_stencil.set_mask(mask, bits)
+
+    def write_fits(self, path: str) -> None:
+        """Write the cutout to a FITS file.
+
+        Parameters
+        ----------
+        path : `str`
+            Local path to the file.
+
+        Notes
+        -----
+        If ``cutout`` is an `Exposure`, this will merge `metadata` into the
+        cutout's attached metadata.  In other cases, `metadata` is written
+        to the primary header without modifying `cutout`.
+        """
+        if isinstance(self.cutout, Exposure):
+            self.cutout.getMetadata().update(self.metadata)
+            self.cutout.writeFits(path)
+        else:
+            self.cutout.writeFits(path, metadata=self.metadata)
+
+
+class ImageCutoutBackend:
+    """High-level interface to the image cutout backend.
+
+    Parameters
+    ----------
+    butler : `Butler`
+        Butler that image cutouts are read from.
+    projection_finder : `ProjectionObject`
+        Object that obtains the WCS and bounding box for butler datasets of
+        different types.  May include caches.
+    output_root : convertible to `ResourcePath`
+        Root of output file URIs.  This will be combined with the originating
+        dataset's UUID and an encoding of the stencil to form the complete URI.
+    filename_template : `str`, optional
+        Template for output file names, not including the ``output_root``
+        prefix or the file type extension.  This is a `str.format` template
+        with named substitutions.  These must include ``stencil`` (the sky
+        stencil fingerprint in hex), and a unique set of fields from ``ref``
+        (the origin `DatasetRef`), such as ``ref.id`` or a combination of the
+        dataset type, `~lsst.daf.butler.CollectionType.RUN`` collection, and
+        data ID, but ``ref.dataId`` may not be expanded.
```

Comment:
If we're not using the Butler to store files, please don't do any of this; it just makes URLs more complicated to no real purpose that I can see. Instead, just write the output into a unique file name (probably via a UUID) and return where it was written. I don't see anything to be gained from encoding the dataset identifier and cutout parameters into the file name.